File: src/main/java/org/example/graph/CourseSchedule.java
package org.example.graph;

import java.util.*;

public class CourseSchedule {

    Map<Integer, List<Integer>> preqMap = new HashMap<>();
    Set<Integer> visit = new HashSet<>();
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        for (int i=0; i < numCourses; i++)
            preqMap.put(i, new ArrayList<>());
        for (int[] prerequisite : prerequisites) {
            preqMap.get(prerequisite[0]).add(prerequisite[1]);
        }

        for (int crs=0; crs < numCourses; crs++) {
            if (!dfs(crs)) return false;
        }

        return true;
     }

    private boolean dfs(int crs) {
        if (visit.contains(crs)) return false;
        if (preqMap.get(crs).isEmpty()) return true;

        visit.add(crs);
        for (int pre: preqMap.get(crs))
            if (!dfs(pre)) return false;
        visit.remove(crs);
        preqMap.put(crs, new ArrayList<>());
        return true;
    }
}

File: src/main/java/org/example/graph/CourseSchedule.py
from typing import List


class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        # map each course to the list of its prerequisites
        preMap = {i: [] for i in range(numCourses)}
        for crs, pre in prerequisites:
            preMap[crs].append(pre)

        # a set to track the nodes visited in the current dfs path
        visit = set()

        def dfs(crs):
            if crs in visit: return False
            if not preMap[crs]: return True

            visit.add(crs)
            for pre in preMap[crs]:
                if not dfs(pre): return False
            visit.remove(crs)
            preMap[crs] = []

            return True

        # we run dfs for each node in case the prereq graph is disjointed
        for crs in range(numCourses):
            if not dfs(crs): return False
        return True
